server/src/main/java/io/druid/segment/realtime/appenderator/Appenderator.java

```diff
@@ -61,8 +61,6 @@
    * This method may trigger a {@link #persistAll(Committer)} using the supplied Committer. If it does this, the
    * Committer is guaranteed to be *created* synchronously with the call to add, but will actually be used
    * asynchronously.
-   * <p>
-   * The add, clear, persist, persistAll, and push methods should all be called from the same thread.
```

Comment:
@gianm Hmm, I missed it. I think synchronization is not enough because we need to guarantee that the execution order matches the method call order. Otherwise, we need a kind of sorted queue which provides such a guarantee. I think the easiest way is enforcing the methods are called by the same thread as before. 

One thing I'm concerned with is changing ```Appenderator.push()``` looks inevitable to implement the same retry mechanism of ```AppenderatorDriver.publish()```. All ```Appenderator.add()```, ```Appenderator.push()```, and ```Appenderator.persist()``` methods should be called by the same thread calling ```AppenderatorDriver```, and to guarantee this in ```AppenderatorDriver.publish()```, ```Appenderator.push()``` should provide the retry mechanism which retries the push until complete because the calling thread of ```push()``` should be ```pushExecutor```. I think it makes sense to add a new push() method which takes a max retry count. 

What do you think?